### PR TITLE
Add support for composite primary keys

### PR DESCRIPTION
**Pull request by warrick-loyaltycorp**
Added support for composite keys in DynamoDB.

DynamoDbModel::find() and DynamoDbModel::delete() in particular required changes to work correctly with a composite primary key.

Also added the DynamoDbModel::compositeKey property. A better approach may be to change the primary key rather than introducing the new compositeKey property. To create a model with a composite primary key:
1. Set the compositeKey property to an array of the attributes in the key.
2. Set the fillable property to the list of non-key attributes
3. When calling find() send an array of key properties with associated values, e.g. find(['id' => 1, 'id2' => 2]).

[Review on Reviewable](https://reviewable.io/reviews/baopham/laravel-dynamodb/5)

### CODE REVIEW

**Comment by baopham** [comment]
wow great! Could you please update the tests and update the README? let me know if you need help with creating a stub db for DynamoDB local.

File: src/DynamoDbModel.php [hunk 1]
```diff
@@ -167,7 +212,24 @@ public static function find($id, array $columns = [])
...
+            }
+            else
+            {
+                $model->id = $id[$model->primaryKey];
+            }
+        }
+        else
+        {
+            $model->id = $id;
+        }
```
Comment (baopham) [inline src/DynamoDbModel.php]: can we wrap all this inside a helper method?

File: src/DynamoDbModel.php [hunk 2]
```diff
@@ -141,14 +165,35 @@ public function delete()
...
+                {
+                    $key[$key_name] = $key_value;
+                }
+            }
+        }
+        else
+        {
+            $key = static::getDynamoDbKey($model, $id);
+        }
+
```
Comment (baopham) [inline src/DynamoDbModel.php]: can we wrap all this inside a helper method? thanks!

File: src/DynamoDbModel.php [hunk 3]
```diff
@@ -128,11 +130,33 @@ public static function create(array $attributes = [])
...
+                {
+                    $key[$key_name] = $key_value;
+                }
+            }
+        }
+        else
+        {
+            $key = static::getDynamoDbKey($this, $this->getKey());
+        }
+
```
Comment (baopham) [inline src/DynamoDbModel.php]: can we wrap all this inside a helper method? thank you :)

**Comment by baopham** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/DynamoDbModel.php, line 156 \[r1\]](https://reviewable.io:443/reviews/baopham/laravel-dynamodb/5#-K8H59Nh5AGxlxK-1W-M-r1-156:-K8H8CSYZzOH8DiKLr6w:-1213433321) ([raw file](https://github.com/baopham/laravel-dynamodb/blob/66150dc6e7c7720e1bed7082989371106b9b3b46/src/DynamoDbModel.php#L156)):*
I also would like that we stick to the PSR-2 as much as we can:

```
if ($expr1) {
    // if body
} elseif ($expr2) {
    // elseif body
} else {
    // else body;
}
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/baopham/laravel-dynamodb/5)*

**Comment by wzedi** [comment]
Updated tests and README.
Also working on support for optimistic locking and timestamps. Will send new pull request for those changes.

I've found I had conflicts when using multiple tables, e.g. If I add the composite key test with a new table and use CompositeTestModel for that table, after instantiating CompositeTestModel and then referring to an instance of TestModel $model->getTable() returns the composite_test_model table causing errors.

I removed the Singleton $instance from DynamoDbModel to avoid this conflict.

**Comment by baopham** [comment]
Thank you! I'll go through this this week. Most likely the weekend.

**Comment by baopham** [comment]
Move to #6

**Pull request closed without merging**